src/riak_kv_wm_stats.erl

```diff
@@ -66,25 +69,67 @@ content_types_provided(ReqData, Context) ->
       {"text/plain", pretty_print}],
      ReqData, Context}.
 
-
 service_available(ReqData, Ctx) ->
     {true, ReqData, Ctx}.
 
+malformed_request(RD, Ctx) ->
+    case wrq:get_qs_value("timeout", RD) of
+        undefined ->
+            {false, RD, Ctx};
+        TimeoutStr ->
+            try
+                case list_to_integer(TimeoutStr) of
+                    Timeout when Timeout > 0 ->
+                        {false, RD, Ctx#ctx{timeout=Timeout}}
+                end
+            catch
+                _:_ ->
+                    {true,
+                        wrq:append_to_resp_body(
+                            io_lib:format(
+                                "Bad timeout value ~0p",
+                                [TimeoutStr]
+                            ),
+                        wrq:set_resp_header(?HEAD_CTYPE, "text/plain", RD)),
+                        Ctx}
+            end
+    end.
+
 forbidden(RD, Ctx) ->
     {riak_kv_wm_utils:is_forbidden(RD), RD, Ctx}.
 
-produce_body(ReqData, Ctx) ->
-    Stats= riak_kv_http_cache:get_stats(),
-    Body = mochijson2:encode({struct, Stats}),
-    {Body, ReqData, Ctx}.
+produce_body(RD, Ctx) ->
+    try 
+        Stats = riak_kv_http_cache:get_stats(Ctx#ctx.timeout),
+        Body = mochijson2:encode({struct, Stats}),
+        {Body, RD, Ctx}
+    catch
+        exit:{timeout, _} ->
+            {
+                {halt, 503},
+                wrq:set_resp_header(
+                    ?HEAD_CTYPE,
+                    "text/plain",
+                    wrq:append_to_response_body(
+                        io_lib:format(
+                            "Request timed out after ~w ms",
+                            [Ctx#ctx.timeout]
+                        ),
+                        RD
+                    )
+                ),
+                Ctx
+            }
+    end.
 
 %% @spec pretty_print(webmachine:wrq(), context()) ->
 %%          {string(), webmachine:wrq(), context()}
 %% @doc Format the respons JSON object is a "pretty-printed" style.
-pretty_print(RD1, C1=#ctx{}) ->
-    {Json, RD2, C2} = produce_body(RD1, C1),
-    {json_pp:print(binary_to_list(list_to_binary(Json))), RD2, C2}.
```

Comment:
Was the reason to do this that Json can be a deeplist and it is flattened in this way to make `json_pp:print` work?

Reply:
Just in case I've reverted it back, as there was presumably a good reason